compiler/rustc_middle/src/ty/context.rs

```diff
@@ -1232,12 +1232,17 @@ impl<'tcx> TyCtxt<'tcx> {
     /// Obtain the given diagnostic item's `DefId`. Use `is_diagnostic_item` if you just want to
     /// compare against another `DefId`, since `is_diagnostic_item` is cheaper.
     pub fn get_diagnostic_item(self, name: Symbol) -> Option<DefId> {
-        self.all_diagnostic_items(()).get(&name).copied()
+        self.all_diagnostic_items(()).name_to_id.get(&name).copied()
+    }
+
+    /// Obtain the diagnostic item's name
+    pub fn get_diagnostic_name(self, id: DefId) -> Option<Symbol> {
+        self.all_diagnostic_items(()).id_to_name.get(&id).copied()
```

Comment:
I believe this is the source of the perf changes. The following should be equivalent I think

```suggestion
        self.diagnostic_items(id.krate).id_to_name.get(&id).copied()
```